rust-symcrypt/src/gcm.rs

```diff
@@ -0,0 +1,377 @@
+        let nonce = hex::decode("cafebabefacedbaddecaf888").unwrap();
+        let auth_data = hex::decode("feedfacedeadbeeffeedfacedeadbeefabaddad2").unwrap();
+        let expected_result = "d9313225f88406e5a55909c5aff5269a86a7a9531534f7da2e4c303d8a318a721c3c0c95956809532fcf0e2449a6b525b16aedf5aa0de657ba637b39";
+        let tag = hex::decode("5bc94fbc3221a5db94fae95ae7121a47").unwrap();
+        let pt = hex::decode("42831ec2217774244b7221b784d0d49ce3aa212f2c02a4e035c17e2329aca12e21d514b25466931c7d8f6a5aac84aa051ba30b396a0aac973d58e091").unwrap();
+
+        let mut gcm_state = GcmState::new(&p_key, &nonce).unwrap();
+        gcm_state.auth(&auth_data);
+        let dst = gcm_state.decrypt_part(&pt);
+        assert_eq!(hex::encode(dst), expected_result);
+        gcm_state.decrypt_final(&tag).unwrap();
+    }
+    #[test]
+    fn test_gcm_decrypt_no_decrypt_part() {
+        let p_key = hex::decode("00000000000000000000000000000000").unwrap();
+        let nonce = hex::decode("000000000000000000000000").unwrap();
+        let auth_data = hex::decode("").unwrap();
+        let tag = hex::decode("58e2fccefa7e3061367f1d57a4e7455a").unwrap();
+
+        let mut gcm_state = GcmState::new(&p_key, &nonce).unwrap();
+        gcm_state.auth(&auth_data);
+        gcm_state.decrypt_final(&tag).unwrap();
+    }
+
+    #[test]
+    fn test_gcm_encrypt_no_encrypt_part() {
+        let p_key = hex::decode("00000000000000000000000000000000").unwrap();
+        let nonce = hex::decode("000000000000000000000000").unwrap();
+        let auth_data = hex::decode("").unwrap();
+
+        let expected_tag = "58e2fccefa7e3061367f1d57a4e7455a";
+
+        let mut gcm_state = GcmState::new(&p_key, &nonce).unwrap();
+        gcm_state.auth(&auth_data);
+        let mut tag: [u8; 16] = [0u8; 16];
+
+        gcm_state.encrypt_final(&mut tag);
+        assert_eq!(hex::encode(tag),expected_tag);
+    } 
+    
+    #[test]
+    fn test_gcm_decrypt_error_message() {
+        let p_key = hex::decode("00000000000000000000000000000000").unwrap();
+        let nonce = hex::decode("000000000000000000000000").unwrap();
+        let auth_data = hex::decode("").unwrap();
+
+        let mut gcm_state = GcmState::new(&p_key, &nonce).unwrap();
+
+        gcm_state.auth(&auth_data);
+        let tag: [u8; 16] = [0u8; 16];
+
+        let result = gcm_state.decrypt_final(&tag);
+        assert_eq!(result.unwrap_err(), SymCryptError::AuthenticationFailure);
+    }
+
+    #[test]
+    fn test_validate_parameters() {
+        let nonce = hex::decode("cafebabefacedbaddecaf888").unwrap();
+        let auth_data = hex::decode("feedfacedeadbeeffeedfacedeadbeefabaddad2").unwrap();
+        let expected_tag = hex::decode("5bc94fbc3221a5db94fae95ae7121a47").unwrap();
+        let pt = hex::decode("d9313225f88406e5a55909c5aff5269a86a7a9531534f7da2e4c303d8a318a721c3c0c95956809532fcf0e2449a6b525b16aedf5aa0de657ba637b39").unwrap();
+
+        unsafe {
```

Comment:
Still thinking of how to wrap this so that is friendly, if you have any ideas please let me know

Reply:
You need to add a rust enum type that is passed to validate_gcm_parameters.